### PR TITLE
Reorganize DBSCAN location and add unit testing

### PR DESCRIPTION
**Pull request by aprokop**
Please make reviewing this a priority, it is holding several future PRs (merging FDBSCAN-Dense and DBSCAN*).

### CODE REVIEW

**Comment by Rombur** [comment]
This PR does not build
```
/var/jenkins/workspace/ArborX_PR-503/test_install/examples/dbscan/dbscan.cpp:13:10: fatal error: 'ArborX_DBSCANVerification.hpp' file not found

#include <ArborX_DBSCANVerification.hpp>

         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

**Comment by aprokop** [comment]
> This PR does not build

@Rombur Thanks. Took me a while to realize that the thing that fails is post-installation build, not the build itself.

I put in the file back in examples, but now test has to reach out to it. Don't particularly like this arrangement, but can't think of a better way.

File: examples/dbscan/ArborX_DBSCANVerification.hpp [hunk 1]
```diff
@@ -115,7 +115,8 @@ bool verifyBoundaryAndNoisePoints(ExecutionSpace const &exec_space,
           if (!is_boundary && labels(i) != -1)
           {
 #ifndef __SYCL_DEVICE_ONLY__
-            printf("Noise point does have index -1: %d [%d]\n", i, labels(i));
+            printf("Noise point does not have index -1: %d [%d]\n", i,
+                   labels(i));
```
Comment (masterleinad) [inline examples/dbscan/ArborX_DBSCANVerification.hpp:119]: The change in formatting is slightly unexpected, but whatever...

File: test/tstDBSCAN.cpp [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+template <typename DeviceType, typename T>
+auto buildView(std::vector<T> const &v)
+{
+  using ExecutionSpace = typename DeviceType::execution_space;
+
+  Kokkos::View<T *, DeviceType> view("Testing::v", v.size());
+  Kokkos::deep_copy(ExecutionSpace{}, view,
+                    Kokkos::View<T const *, Kokkos::HostSpace,
+                                 Kokkos::MemoryTraits<Kokkos::Unmanaged>>(
+                        v.data(), v.size()));
```
Comment (masterleinad) [inline test/tstDBSCAN.cpp:26]:
```suggestion
  Kokkos::deep_copy(view,
                    Kokkos::View<T const *, Kokkos::HostSpace,
                                 Kokkos::MemoryTraits<Kokkos::Unmanaged>>(
                        v.data(), v.size()));
```
Using the three-argument deep-copy here is a little pointless if we don't pass an execution space instance as argument. We are relying on the fact that we always get the same stream when calling the default constructor.

File: test/tstDBSCAN.cpp [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+    BOOST_TEST(verifyDBSCAN(space, points, 3 * r, 3,
+                            buildView<DeviceType, int>({5, 5, 5, 5})));
+    BOOST_TEST(verifyDBSCAN(space, points, 3 * r, 4,
+                            buildView<DeviceType, int>({7, 7, 7, 7})));
+    BOOST_TEST(verifyDBSCAN(space, points, 3 * r, 5,
+                            buildView<DeviceType, int>({-1, -1, -1, -1})));
+  }
+
+  {
+    // check for briging effect
```
Comment (masterleinad) [inline test/tstDBSCAN.cpp:88]:
```suggestion
    // check for bridging effect
```
?

File: test/tstDBSCAN.cpp [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+                            buildView<DeviceType, int>({3, 3, 3, -1})));
+    BOOST_TEST(verifyDBSCAN(space, points, 2 * r, 4,
+                            buildView<DeviceType, int>({-1, -1, -1, -1})));
+
+    BOOST_TEST(verifyDBSCAN(space, points, 3 * r, 2,
+                            buildView<DeviceType, int>({5, 5, 5, 5})));
+    BOOST_TEST(verifyDBSCAN(space, points, 3 * r, 3,
+                            buildView<DeviceType, int>({5, 5, 5, 5})));
+    BOOST_TEST(verifyDBSCAN(space, points, 3 * r, 4,
+                            buildView<DeviceType, int>({7, 7, 7, 7})));
```
Comment (masterleinad) [inline test/tstDBSCAN.cpp:82]: Why are the labels different for these two cases?

Comment (aprokop) [reply test/tstDBSCAN.cpp:82]: The labels don't matter per say. What matters is that the points in a cluster have the same ones. Thus, I'm using some random labels to make sure that we don't rely on any particular one.

File: examples/dbscan/ArborX_DBSCANVerification.hpp (hunk 1 above)
Comment (aprokop) [reply examples/dbscan/ArborX_DBSCANVerification.hpp:119]: 🤷

File: test/tstDBSCAN.cpp [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+    BOOST_TEST(
+        verifyDBSCAN(space, points, 3 * r, 3, dbscan(space, points, 3 * r, 3)));
+    BOOST_TEST(
+        verifyDBSCAN(space, points, 3 * r, 4, dbscan(space, points, 3 * r, 4)));
+    BOOST_TEST(
+        verifyDBSCAN(space, points, 3 * r, 5, dbscan(space, points, 3 * r, 5)));
+  }
+
+  {
+    // check for briging effect
```
Comment (aprokop) [inline test/tstDBSCAN.cpp:157]:
```suggestion
    // check for bridging effect
```

File: test/tstDBSCAN.cpp (hunk 4 above)
Comment (masterleinad) [reply test/tstDBSCAN.cpp:82]: Ah, OK. That's what I thought.

**Comment by Rombur** [comment]
The PR does not build with CUDA Clang. The compiler chokes on the new test.

**Comment by aprokop** [comment]
> The PR does not build with CUDA Clang. The compiler chokes on the new test.

I noticed that too, but so far have no clue as to how to fix that.

**Comment by Rombur** [comment]
Is it because you didn't define `#define BOOST_TEST_MODULE ` (see [here](https://github.com/arborx/ArborX/blob/2df522a5ee13a99771cab4a4dafb7e99ed7d08fc/test/tstDetailsCrsGraphWrapperImpl.cpp#L20)) or `BOOST_AUTO_TEST_SUITE` (see [here](https://github.com/arborx/ArborX/blob/2df522a5ee13a99771cab4a4dafb7e99ed7d08fc/test/tstQueryTreeDegenerate.cpp#L21)). Not sure that's the problem since the other builds work but the macros definition in boost is a mess because each compiler does something different.

**Comment by aprokop** [comment]
> Is it because you didn't define `#define BOOST_TEST_MODULE ` (see [here](https://github.com/arborx/ArborX/blob/2df522a5ee13a99771cab4a4dafb7e99ed7d08fc/test/tstDetailsCrsGraphWrapperImpl.cpp#L20)) or `BOOST_AUTO_TEST_SUITE` (see [here](https://github.com/arborx/ArborX/blob/2df522a5ee13a99771cab4a4dafb7e99ed7d08fc/test/tstQueryTreeDegenerate.cpp#L21)). Not sure that's the problem since the other builds work but the macros definition in boost is a mess because each compiler does something different.

I guess not, it has the same failure even with `BOOST_AUTO_TEST_SUITE` defined.

**Review by dalg24 — COMMENTED** [review]
`#include "BoostTest_CUDA_clang_workarounds.hpp"`

I am NOT done reviewing

File: examples/dbscan/ArborX_DBSCANVerification.hpp (hunk 1 above)
Comment (aprokop) [reply examples/dbscan/ArborX_DBSCANVerification.hpp:119]: > How is it unexpected?

@dalg24 It means that a real noise point has index different from -1, which means it was clustered, which should not have happened.

File: test/CMakeLists.txt [hunk 6]
```diff
@@ -132,6 +132,12 @@ target_compile_definitions(ArborX_DetailsCrsGraphWrapperImpl.exe PRIVATE BOOST_T
 target_include_directories(ArborX_DetailsCrsGraphWrapperImpl.exe PRIVATE ${CMAKE_CURRENT_BINARY_DIR})
 add_test(NAME ArborX_DetailsCrsGraphWrapperImpl_Test COMMAND ./ArborX_DetailsCrsGraphWrapperImpl.exe)
 
+add_executable(ArborX_Clustering.exe tstDBSCAN.cpp utf_main.cpp)
+target_link_libraries(ArborX_Clustering.exe PRIVATE ArborX Boost::unit_test_framework)
+target_compile_definitions(ArborX_Clustering.exe PRIVATE BOOST_TEST_DYN_LINK)
+target_include_directories(ArborX_Clustering.exe PRIVATE ${CMAKE_CURRENT_BINARY_DIR} ${CMAKE_SOURCE_DIR}/examples/dbscan)
```
[the remark this answers is not in the document]
Comment (aprokop) [reply test/CMakeLists.txt:138]: `ArborX_EnableDeviceTypes.hpp`

**Comment by aprokop** [comment]
> `#include "BoostTest_CUDA_clang_workarounds.hpp"`

@dalg24 Thanks! That fixed it.

**Review by dalg24 — COMMENTED** [review]
I am not sure `docs/` is the right place for the ECL license.
I don't have any major objection.  I would like to organize the testing slightly differently but that can be fixed in later PRs.

**Comment by aprokop** [comment]
retest this please

**Comment by aprokop** [comment]
Does running "retest this please" not trigger tests if the last commit has "[skip ci]"?

In any case, this passes all tests, and the last commit should have fixed the warning, so the PR may be merged as is as soon as someone approves it.

**Review by dalg24 — APPROVED** [review]
> Realistically, right now docs/ is just a potpourri of random things.

Not true.  This PR is a step in that direction though.

**Pull request merged**